web_poet/mixins.py

```diff
@@ -78,3 +81,30 @@ def urljoin(self, url: str) -> str:
         url and baseurl of the response"""
         # FIXME: move it to HttpResponse
         return urljoin(self.base_url, url)
+
+
+class UrlMixin(abc.ABC):
+    """Common shortcut methods for working with the URL of HTTP requests and
+    responses.
+
+    This mixin could be used with Page Object base classes.
+
+    It requires the ``url`` attribute to be present.
+    """
+
+    _cached_base_url = None
+    url: Union[str, RequestUrl, ResponseUrl]
+
+    @property
+    def _base_url(self) -> str:
+        if self._cached_base_url is None:
+            text = getattr(self, "text", "")[:4096]
+            self._cached_base_url = get_base_url(text, str(self.url))
+        return self._cached_base_url
+
+    def urljoin(self, url: Union[str, RequestUrl, ResponseUrl]) -> RequestUrl:
```

Comment:
Is `RequestUrl` the right choice for the output? Should it instead match the input type? Should it always be `str`?

Reply:
A good question.

RequestUrl and str seems reasonable choices; I don't see a use case for matching the input type.

RequestUrl looks like a right choice, but we need to make sure that it works well with zyte-common-items. For example, it must be possible to extract image or other URLs using .urljoin, and create items with the result, without boilerplate.